src/jdk.compiler/share/classes/com/sun/tools/javac/launcher/Main.java

```diff
@@ -412,32 +415,57 @@ private String compile(Path file, List<String> javacOpts, Context context) throw
      * will load recently compiled classes from memory.
      *
      * @param mainClassName the class to be executed
-     * @param appArgs the arguments for the {@code main} method
+     * @param mainArgs the arguments for the {@code main} method
      * @param context the context for the class to be executed
      * @throws Fault if there is a problem finding or invoking the {@code main} method
      * @throws InvocationTargetException if the {@code main} method throws an exception
      */
-    private void execute(String mainClassName, String[] appArgs, Context context)
+    private void execute(String mainClassName, String[] mainArgs, Context context)
             throws Fault, InvocationTargetException {
         System.setProperty("jdk.launcher.sourcefile", context.file.toString());
         ClassLoader cl = context.getClassLoader(ClassLoader.getSystemClassLoader());
         try {
             Class<?> appClass = Class.forName(mainClassName, true, cl);
-            Method main = appClass.getDeclaredMethod("main", String[].class);
-            int PUBLIC_STATIC = Modifier.PUBLIC | Modifier.STATIC;
-            if ((main.getModifiers() & PUBLIC_STATIC) != PUBLIC_STATIC) {
+            Method mainMethod = MainMethodFinder.findMainMethod(appClass);
+            int mods = mainMethod.getModifiers();
+            boolean isStatic = Modifier.isStatic(mods);
+            boolean isPublic = Modifier.isPublic(mods);
+            boolean noArgs = mainMethod.getParameterCount() == 0;
+
+            if (!PreviewFeatures.isEnabled() && (!isStatic || !isPublic)) {
                 throw new Fault(Errors.MainNotPublicStatic);
             }
-            if (!main.getReturnType().equals(void.class)) {
+
+            if (!mainMethod.getReturnType().equals(void.class)) {
                 throw new Fault(Errors.MainNotVoid);
             }
-            main.setAccessible(true);
-            main.invoke(0, (Object) appArgs);
+
+            // Similar to sun.launcher.LauncherHelper#executeMainClass
+            // but duplicated here to prevent additional launcher frames
+            mainMethod.setAccessible(true);
+
+            if (isStatic) {
+                if (noArgs) {
+                    mainMethod.invoke(appClass);
+                } else {
+                    mainMethod.invoke(appClass, (Object)mainArgs);
+                }
+            } else {
+                Constructor<?> constructor = appClass.getDeclaredConstructor();
+                constructor.setAccessible(true);
+                Object instance = constructor.newInstance();
+
+                if (noArgs) {
+                    mainMethod.invoke(instance);
+                } else {
+                    mainMethod.invoke(instance, (Object)mainArgs);
+                }
+            }
         } catch (ClassNotFoundException e) {
             throw new Fault(Errors.CantFindClass(mainClassName));
         } catch (NoSuchMethodException e) {
```

Comment:
`getDeclaredConstructor` can also throw a NSME but this error message would still say that the main method couldn't be found unless at this point we are sure that the class must have a no-args constructor

Reply:
I was avoiding moving things so not to change behaviour when not in preview, but I think I really need to break it down into separate try blocks; get method, get constructor, execute method. Let me attempt.